src/content_script/index.ts

```diff
@@ -37,23 +38,23 @@ const injectScript = () => {
 const setupCommunication = () => {
   const port = browser.runtime.connect(null, {
     // TODO: change name to window.location.hostname
-    name: CONTENT_SCRIPT_PORT_NAME
+    name: `${MessageOrigin.PAGE}/${v4()}`
   });
 
-  port.onMessage.addListener(data => {
-    logger.debug('message from ext', data);
-    window.postMessage(data, window.origin);
-  });
+  const portStream = new RuntimePortDuplex(port);
+  const inpageWriteStream = new MultiplexWindowMessageDuplex(window,MessageOrigin.CONTENT, MessageOrigin.PAGE);
+  portStream.pipe(inpageWriteStream).pipe(addOrigin(port.name)).pipe(portStream);
 
-  fromEvent<MessageEvent>(window, 'message')
-    .pipe(filter(e => e.source === window))
-    .subscribe(e => port.postMessage(e.data));
+  return {
+    inpageWriteStream,
+    portStream
+  };
 };
 
-const init = () => {
-  window.postMessage({ type: 'init' }, window.origin);
+const init = (inpageWriteStream: RuntimePortDuplex) => {
```

Comment:
just notice, this place we call `init(portStream);`, so should we use portStream but inpageWriteStream as the name of the parameter here